pkg/controller/certificatesigningrequests/venafi/venafi.go

```diff
@@ -0,0 +1,206 @@
+		return err
+	}
+
+	var customFields []venafiapi.CustomField
+	if annotation, exists := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey]; exists && annotation != "" {
+		err := json.Unmarshal([]byte(annotation), &customFields)
+		if err != nil {
+			message := fmt.Sprintf("Failed to parse %q annotation: %s", experimentalapi.CertificateSigningRequestVenafiCustomFieldsAnnotationKey, err)
+			v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", message)
+			util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", message)
+			_, err = v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+			return err
+		}
+	}
+
+	duration, err := pki.DurationFromCertificateSigningRequest(csr)
+	if err != nil {
+		message := fmt.Sprintf("Failed to parse requested duration: %s", err)
+		log.Error(err, message)
+		v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorParseDuration", message)
+		util.CertificateSigningRequestSetFailed(csr, "ErrorParseDuration", message)
+		_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	// The signing process with Venafi is slow. The "pickupID" allows us to track
+	// the progress of the certificate signing. It is set as an annotation the
+	// first time the Certificate is reconciled.
+	pickupID := csr.GetAnnotations()[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey]
+
+	// check if the pickup ID annotation is there, if not set it up.
+	if len(pickupID) == 0 {
+		pickupID, err := client.RequestCertificate(csr.Spec.Request, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
+
+			case venaficlient.ErrCustomFieldsType:
+				log.Error(err, "")
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorCustomFields", err.Error())
+				util.CertificateSigningRequestSetFailed(csr, "ErrorCustomFields", err.Error())
+				_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+				return err
+
+			default:
+				message := fmt.Sprintf("Failed to request venafi certificate: %s", err)
+				log.Error(err, message)
+				v.recorder.Event(csr, corev1.EventTypeWarning, "ErrorRequest", message)
+				util.CertificateSigningRequestSetFailed(csr, "ErrorRequest", message)
+				_, err := v.certClient.UpdateStatus(ctx, csr, metav1.UpdateOptions{})
+				return err
+			}
+		}
+
+		if csr.Annotations == nil {
+			csr.Annotations = make(map[string]string)
+		}
+		csr.Annotations[experimentalapi.CertificateSigningRequestVenafiPickupIDAnnotationKey] = pickupID
+		_, err = v.certClient.Update(ctx, csr, metav1.UpdateOptions{})
+		return err
+	}
+
+	certPem, err := client.RetrieveCertificate(pickupID, csr.Spec.Request, duration, customFields)
+	if err != nil {
+		switch err.(type) {
+		case endpoint.ErrCertificatePending, endpoint.ErrRetrieveCertificateTimeout:
+			message := "Venafi certificate still in a pending state, the request will be retried"
```

Comment:
nit: by "request", do you mean the action of polling TPP to know whether the certificate is done being signed?

```suggestion
			message := "Venafi certificate still in a pending state, waiting"
```